### ISSUE TITLE
Can't find the expression or hash "ha2dimzxguzto" in the Drab.Live.Cache

### ISSUE THREAD

**Issue by GBH** [grych/drab#37]
Works perfectly fine in development. Deployed to production and it's crashing. Any clue what that might be about?

Here's stack trace:

```
23:20:46.844 request_id=tj62j2os7fdkg3lu8h5vurldr0mq1ol2 [info] GET /tours/1-can-am-tour-of-2016
23:20:46.899 request_id=tj62j2os7fdkg3lu8h5vurldr0mq1ol2 [info] Sent 200 in 55ms 
23:20:47.462 [info] JOIN "__drab:same_path:/tours/1-can-am-tour-of-2016" to Drab.Channel
  Transport:  Phoenix.Transports.WebSocket (2.0.0)
  Serializer:  Phoenix.Transports.V2.WebSocketSerializer
  Parameters: %{}
23:20:47.463 [info] Replied __drab:same_path:/tours/1-can-am-tour-of-2016 :ok
23:20:49.726 [error] Drab Handler failed with the following exception:
** (RuntimeError) Can't find the expression or hash "ha2dimzxguzto" in the Drab.Live.Cache
    (drab) lib/drab/live/cache.ex:40: Drab.Live.Cache.get/1
    (drab) lib/drab/live.ex:298: anonymous fn/6 in Drab.Live.do_poke/5
    (elixir) lib/enum.ex:1755: Enum."-reduce/3-lists^foldl/2-0-"/3
    (drab) lib/drab/live.ex:295: Drab.Live.do_poke/5
    (drab) lib/drab.ex:249: anonymous fn/6 in Drab.handle_event/6


23:20:54.729 [error] Error in process #PID<0.2708.0> on node :"loaded_bike@127.0.0.1" with exit value:
{{:badmatch, {:error, "timed out after 5000 ms."}},
 [{Drab, :failed, 2, [file: 'lib/drab.ex', line: 286]},
  {Drab, :"-handle_event/6-fun-3-", 6, [file: 'lib/drab.ex', line: 258]}]}

23:20:54.729 [error] Drab Process #PID<0.2708.0> died because of {:badmatch, {:error, "timed out after 5000 ms."}}
    (drab) lib/drab.ex:286: Drab.failed/2
    (drab) lib/drab.ex:258: anonymous fn/6 in Drab.handle_event/6
```
EDIT:

I dug around code a bit and found this: https://github.com/grych/drab/blob/master/lib/drab/live/cache.ex#L37

When running that line on the build server (same as prod) I get expected structure (template as a function thing). I don't even know how that method would produce output other than `[{k: v}]` or `[]`

Is there any other way to debug this in running production?

EDIT 2:
```
iex(loaded_bike@127.0.0.1)7> :dets.info("priv/hashes_expressions.drab.cache.prod")
[type: :set, keypos: 1, size: 0, file_size: 5464,
 filename: "priv/hashes_expressions.drab.cache.prod"]

iex(loaded_bike@127.0.0.1)11> :dets.lookup("priv/hashes_expressions.drab.cache.prod", "ha2dimzxguzto")
{:error, {:file_error, "priv/hashes_expressions.drab.cache.prod", :enoent}}
```

Huh. Cache file is there but has no data in it?

Locally, however:

```
iex(2)> :dets.info("priv/hashes_expressions.drab.cache.prod")                   
[type: :set, keypos: 1, size: 13, file_size: 34266,
 filename: "priv/hashes_expressions.drab.cache.prod"]

iex(1)> :dets.lookup("priv/hashes_expressions.drab.cache.prod", "ha2dimzxguzto")
[{"ha2dimzxguzto",
  {:expr, ...
```
I triple-checked that same file is locally and one deployed:

live:
```
-rw-rw-r-- 1 deploy deploy 34266 Aug 10 00:47 hashes_expressions.drab.cache.prod
```

local:
```
-rw-r--r--  1 oleg oleg 34266 Aug  9 17:42 hashes_expressions.drab.cache.prod
```

EDIT 3:

Ok, now I'm really confused. I thought maybe I'm looking at the wrong thing and there's 5464 byte cache file somewhere on production. Nopes:

```
deploy@loaded-bike-app:~/loaded_bike$ find ~/ -name hashes_expressions.drab.cache.prod | xargs du -sh
36K	/home/deploy/builds/priv/hashes_expressions.drab.cache.prod
36K	/home/deploy/builds/_build/prod/lib/loaded_bike/priv/hashes_expressions.drab.cache.prod
36K	/home/deploy/builds/_build/prod/rel/loaded_bike/lib/loaded_bike-0.0.1+205-f804a9b/priv/hashes_expressions.drab.cache.prod
36K	/home/deploy/loaded_bike/lib/loaded_bike-0.0.1+205-f804a9b/priv/hashes_expressions.drab.cache.prod
```

EDIT 4:

Ah, so prod doesn't even see the file in the first place (local returns `true`)

```
iex(loaded_bike@127.0.0.1)6> :dets.is_dets_file("priv/hashes_expressions.drab.cache.prod") 
{:error, {:file_error, "priv/hashes_expressions.drab.cache.prod", :enoent}}
```

EDIT 5:

Ok, I figured it out. App was looking to the cache file in the non-existant folder:

```
~/loaded_bike/priv
```
and not in where the lib actually is:
```
~/loaded_bike/lib/loaded_bike-0.0.1+205-f804a9b/priv
```

Seems like a configuration issue. But there's no issues serving files from `priv/static` ???

**Comment by grych** [comment]
The cache is created during the compile time, and using during the runtime. I think your production environment uses different path for those.

The issue is probably how Drab opens the file, now it is just simple "priv/filename". I am going to use the full path, I hope that would solve it.

What service provider do you use for production? I want to reproduce it myself.

**Comment by OvermindDL1** [comment]
https://github.com/grych/drab/blob/63ebb803c28d268681cf775fe9d235076f388a89/lib/drab/live/cache.ex#L9
```elixir
  @cache_file "priv/hashes_expressions.drab.cache"
```

*Entirely* wrong!

Never ever reference the `priv` directory directly!  Every application has its own priv directory and they can be moved and shuffled and even renamed with impunity by the runtime system!!!

To get the priv directory always run `:code.priv_dir(<application_name>)`, like:
```elixir
iex> :code.priv_dir(:my_application)
'/home/my_user/my_server/_build/prod/rel/my_server/lib/my_server-0.0.3/priv'
```
Notice it returns a charlist, just to note.  And it will always be updated with the correct value by the erlang runtime system (ERTS).

You should have no direct references to a `priv` directory *anywhere* at all after a `mix release` has been run.

I've not used Drab.Live yet otherwise I would have noticed this (old project that is being slowly migrated to drab, I'm using some other parts of Drab quite a bit though)...  ^.^;

**Comment by GBH** [comment]
Prod is on DigitalOcean and I deploy there using edeliver/distillery. Repo is actually public if you want to try it out: https://github.com/GBH/loaded.bike

Love this project btw. Coming from Rails it feels like black magic.

**Comment by grych** [comment]
Thank you both. There is more usage of priv than `Drab.Live`, I'll fix it.

**Comment by grych** [comment]
Fixed in v0.5.4, please re-check and re-open in case

**Comment by GBH** [comment]
Confirming as fixed.
